src/main/java/org/openelisglobal/common/rest/provider/NonConformEventsRestController.java

```diff
@@ -0,0 +1,250 @@
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class NonConformEventsRestController {
+
+  private final SampleService sampleService;
+  private final SampleItemService sampleItemService;
+  private final SearchResultsService searchResultsService;
+  private final NonConformingEventWorker nonConformingEventWorker;
+  private final NceCategoryService nceCategoryService;
+  private final RequesterService requesterService;
+
+  private static final String USER_SESSION_DATA = "userSessionData";
+
+  @Autowired
+  private SystemUserService systemUserService;
+
+  public NonConformEventsRestController(
+    SampleService sampleService,
+    SampleItemService sampleItemService,
+    SearchResultsService searchResultsService,
+    NonConformingEventWorker nonConformingEventWorker,
+    NceCategoryService nceCategoryService,
+    RequesterService requesterService
+  ) {
+    this.sampleService = sampleService;
+    this.sampleItemService = sampleItemService;
+    this.searchResultsService = searchResultsService;
+    this.nonConformingEventWorker = nonConformingEventWorker;
+    this.nceCategoryService = nceCategoryService;
+    this.requesterService = requesterService;
+  }
+
+  @GetMapping(
+    value = "/rest/nonconformevents",
+    produces = MediaType.APPLICATION_JSON_VALUE
+  )
+  public ResponseEntity<?> getNCESampleSearch(
+    @RequestParam(required = false) String lastName,
+    @RequestParam(required = false) String firstName,
+    @RequestParam(required = false) String STNumber,
+    @RequestParam(required = false) String labNumber
+  ) {
+    try {
+      List<Sample> searchResults;
+      if (labNumber != null) {
+        Sample sample = sampleService.getSampleByAccessionNumber(labNumber);
+        searchResults = sample != null ? List.of(sample) : List.of();
+      } else {
+        List<PatientSearchResults> results =
+          searchResultsService.getSearchResults(
+            lastName,
+            firstName,
+            STNumber,
+            "",
+            "",
+            "",
+            "",
+            "",
+            "",
+            ""
+          );
+        searchResults = results
+          .stream()
+          .flatMap(
+            patientSearchResults ->
+              sampleService
+                .getSamplesForPatient(patientSearchResults.getPatientID())
+                .stream()
+          )
+          .collect(Collectors.toList());
+      }
+
+      if (searchResults.isEmpty()) {
+        return ResponseEntity.notFound().build();
+      } else {
+        List<Map<String, Object>> temp = new ArrayList<>();
+        for (Sample sample : searchResults) {
+          temp.add(addSample(sample));
+        }
+        return ResponseEntity.ok().body(Map.of("results", temp));
```

Comment:
generally we dont want to work with loosely typed objexts. can you create a Java Class with stronly typed properties to represent this map ??